File: app/src/main/java/net/fionix/minima/WidgetFactory.java
package net.fionix.minima;

import android.content.Context;
import android.content.Intent;
import android.widget.RemoteViews;
import android.widget.RemoteViewsService;

import java.util.ArrayList;

public class WidgetFactory implements RemoteViewsService.RemoteViewsFactory {

    public Context context;
    public int resourceId;
    ArrayList<ClassTimetable> arrayListTimetable;

    public WidgetFactory(Context context, Intent intent) {
        this.context = context;
        resourceId = Integer.valueOf(intent.getData().getSchemeSpecificPart()) - WidgetProvider.randomNumber;
        arrayListTimetable = ClassMinima.filterForToday(ClassMinima.loadTimetable(context));
    }

    @Override
    public void onCreate() {

    }

    @Override
    public void onDataSetChanged() {

    }

    @Override
    public void onDestroy() {

    }

    @Override
    public int getCount() {
        if (arrayListTimetable != null) {
            return arrayListTimetable.size();
        } else return 0;
    }

    @Override
    public RemoteViews getViewAt(int i) {
        RemoteViews row = new RemoteViews(context.getPackageName(), R.layout.listview_widget);

        String courseName = "";

        ArrayList<ClassCourse> arrayListCourseCode = ClassMinima.loadCourse(context);
        for (int j = 0; j < arrayListCourseCode.size(); j++) {
            if (arrayListCourseCode.get(j).code.contains(arrayListTimetable.get(i).course)) {
                if (arrayListCourseCode.get(j).title.isEmpty())
                    courseName = "You can edit course name at the Course tab";
                else courseName = arrayListCourseCode.get(j).title;
                break;
            }
        }

        row.setTextViewText(R.id.textViewCourseName, courseName);
        row.setTextViewText(R.id.textViewCourseCode, arrayListTimetable.get(i).course);
        row.setTextViewText(R.id.textViewGroup, arrayListTimetable.get(i).group);
        row.setTextViewText(R.id.textViewTime, arrayListTimetable.get(i).start + " - " + arrayListTimetable.get(i).end);
        row.setTextViewText(R.id.textViewLocation, arrayListTimetable.get(i).location.toUpperCase());
        return row;
    }

    @Override
    public RemoteViews getLoadingView() {
        return null;
    }

    @Override
    public int getViewTypeCount() {
        return 1;
    }

    @Override
    public long getItemId(int i) {
        return 0;
    }

    @Override
    public boolean hasStableIds() {
        return false;
    }
}

File: app/src/main/java/net/fionix/minima/widget/WidgetFactory.kt
package net.fionix.minima.widget

import android.appwidget.AppWidgetManager
import android.content.Context
import android.content.Intent
import android.widget.RemoteViews
import android.widget.RemoteViewsService.RemoteViewsFactory
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.GlobalScope
import kotlinx.coroutines.launch
import net.fionix.minima.R
import net.fionix.minima.database.DatabaseMain
import net.fionix.minima.model.EntityTimetable
import java.text.SimpleDateFormat
import java.util.*
import kotlin.collections.ArrayList

class WidgetFactory(var context: Context, intent: Intent) : RemoteViewsFactory {

    private val appWidgetId = intent.getIntExtra(AppWidgetManager.EXTRA_APPWIDGET_ID, AppWidgetManager.INVALID_APPWIDGET_ID) - WidgetReceiver.randomNumber
    private lateinit var timetableList: ArrayList<EntityTimetable>

    override fun getViewAt(index: Int): RemoteViews {

        // view
        val removeViews = RemoteViews(context.packageName, R.layout.list_three_line_linear)

        // title
        var courseCode: String = timetableList[index].courseCode
        if (timetableList[index].courseName.isNotEmpty()) {
            courseCode = courseCode + " - " + timetableList[index].courseName
        }
        removeViews.setTextViewText(R.id.title_text, courseCode)

        // secondary
        val secondaryString: String = timetableList[index].timetableDay + ", " + timetableList[index].timetableTimeStart + " - " + timetableList[index].timetableTimeEnd
        removeViews.setTextViewText(R.id.secondary_text, secondaryString)

        // tertiary
        removeViews.setTextViewText(R.id.tertiary_text, timetableList[index].timetableVenue)

        // return
        return removeViews
    }

    override fun getLoadingView(): RemoteViews? {
        return null
    }

    override fun getCount(): Int {
        if (!this::timetableList.isInitialized) {
            return 0
        }
        return timetableList.size
    }

    override fun getItemId(index: Int): Long {
        return 0
    }

    override fun getViewTypeCount(): Int {
        return 1
    }

    override fun hasStableIds(): Boolean {
        return false
    }

    override fun onCreate() {
        GlobalScope.launch(Dispatchers.IO) {
            timetableList = ArrayList(DatabaseMain.getDatabase(context).timetableDao().getListByDay(SimpleDateFormat("EEEE", Locale.ENGLISH).format(Calendar.getInstance().time)))
        }
    }

    override fun onDataSetChanged() {}

    override fun onDestroy() {}

}